examples/test_flow_dir/test_flow_dir.c

```diff
@@ -151,32 +163,103 @@ do_stats_display(struct rte_mbuf *pkt) {
         }
 }
 
+/*
+ * This function populates a set of flows to the sdn flow table.
+ * For each new flow a service chain is created and appended.
+ * Each service chain is of max 4 length. This function is not enabled by default.
+ * Users may update the predefined struct with their own rules here.
+ */
+
+static void
+populate_sample_ipv4(void) {
+        struct onvm_flow_entry *flow_entry = (struct onvm_flow_entry *)
+                rte_calloc(NULL, 1, sizeof(struct onvm_flow_entry), 0);
+        if (flow_entry == NULL) {
+                rte_exit(EXIT_FAILURE, "Unable to populate flow.\n");
+        }
+        struct test_add_key {
+                struct onvm_ft_ipv4_5tuple key;
+                uint8_t destinations[ONVM_MAX_CHAIN_LENGTH];
+        };
+
+        struct test_add_key keys[] = {
+                {{RTE_IPV4(100, 10, 0, 0), RTE_IPV4(100, 10, 0, 0),  1, 1, IPPROTO_TCP}, {2,3}},
+                {{RTE_IPV4(102, 10, 0, 0), RTE_IPV4(101, 10, 0, 1),  0, 1, IPPROTO_TCP}, {4,3,2,1}},
+                {{RTE_IPV4(103, 10, 0, 0), RTE_IPV4(102, 10, 0, 1),  0, 1, IPPROTO_TCP}, {2,1}},
+                {{RTE_IPV4(10, 11, 1, 17), RTE_IPV4(10, 11, 1, 17),  1234, 1234, IPPROTO_UDP}, {4,3,2}},
+        };
+
+        uint32_t num_keys = RTE_DIM(keys);
+        for (uint32_t i = 0; i < num_keys; i++) {
+                struct onvm_ft_ipv4_5tuple key;
+                key = keys[i].key;
+                int ret = onvm_ft_lookup_key(sdn_ft, &key, (char **)&flow_entry);
+                if (ret >= 0) {
+                        continue;
+                } else {
+                        printf("\nAdding Key: ");
+                        _onvm_ft_print_key(&key);
+                        ret = onvm_ft_add_key(sdn_ft, &key, (char **)&flow_entry);
+                        if (ret < 0) {
+                                printf("Unable to add key.");
+                                continue;
+                        }
+                        printf("Creating new service chain.\n");
+                        flow_entry->sc = onvm_sc_create();
+                        uint32_t num_dest = RTE_DIM(keys[i].destinations);
+                        for (uint32_t j = 0; j < num_dest; j++) {
+                                if (keys[i].destinations[j] == 0) {
+                                        continue;
+                                }
+                                printf("Appending Destination: %d \n", keys[i].destinations[j]);
+                                onvm_sc_append_entry(flow_entry->sc, ONVM_NF_ACTION_TONF, keys[i].destinations[j]);
+                        }
+                }
+        }
+        flow_entry = NULL;
+        rte_free(flow_entry);
```

Comment:
Am I missing something here or should these be flipped?

Reply:
Thank You.